src/dispatch/incident_cost/service.py

```diff
@@ -104,14 +112,95 @@ def get_engagement_multiplier(participant_role: str):
+    return incident_review_prep + incident_review_meeting
+
+
+def calculate_incident_response_cost_with_cost_model(
+    incident: Incident, db_session: SessionLocal
+) -> int:
+    """Calculates the cost of an incident using the incident's cost model."""
 
     participants_total_response_time_seconds = 0
 
+    # Get the cost model. Iterate through all the listed activities we want to record.
+    for activity in incident.incident_cost_model.activities:
+        plugin_instance = plugin_service.get_active_instance_by_slug(
+            db_session=db_session,
+            slug=activity.plugin_event.plugin.slug,
+            project_id=incident.project.id,
+        )
+        if not plugin_instance:
+            log.warning(
+                f"Cannot fetch cost model activity. Its associated plugin {activity.plugin_event.plugin.title} is not enabled."
+            )
+            continue
+
+        oldest = "0"
+        response_cost_type = incident_cost_type_service.get_default(
+            db_session=db_session, project_id=incident.project.id
+        )
+        incident_response_cost = get_by_incident_id_and_incident_cost_type_id(
+            db_session=db_session,
+            incident_id=incident.id,
+            incident_cost_type_id=response_cost_type.id,
+        )
+        if incident_response_cost:
+            oldest = incident_response_cost.updated_at.replace(tzinfo=timezone.utc).timestamp()
+
+        # Array of sorted (timestamp, user_id) tuples.
+        incident_events = plugin_instance.instance.fetch_incident_events(
+            db_session=db_session,
+            subject=incident,
+            plugin_event_id=activity.plugin_event.id,
+            oldest=oldest,
+        )
+
+        for ts, user_id in incident_events:
+            participant = participant_service.get_by_incident_id_and_conversation_id(
+                db_session=db_session,
+                incident_id=incident.id,
+                user_conversation_id=user_id,
+            )
+            if not participant:
+                log.warning("Cannot resolve participant.")
+                continue
+
+            activity_in = IncidentParticipantActivityCreate(
+                plugin_event=activity.plugin_event,
+                started_at=ts,
+                ended_at=ts + timedelta(seconds=activity.response_time_seconds),
+                participant=ParticipantRead(id=participant.id),
+                incident=incident,
+            )
+
+            if participant_response_time := incident_participant_activity_service.create_or_update(
+                db_session=db_session, activity_in=activity_in
+            ):
+                participants_total_response_time_seconds += (
+                    participant_response_time.total_seconds()
+                )
+
+    # Calculate and round up the hourly rate.
+    hourly_rate = math.ceil(
+        incident.project.annual_employee_cost / incident.project.business_year_hours
+    )
```

Comment:
Should this be pushed down to the incident cost model, so we can have multiple cost models with different values for employee cost and biz year hours?

Reply:
Just to verify the intention, by adding these fields to the incident cost model, we'd be able to more closely estimate response costs based on adjustable employee hourly rates?

The reason I've kept it project-level is because:
 - I'd like to eventually include the option to retrieve an individual's (or their function's average) hourly rate for more accurate cost estimations.
 - If that data is unavailable, incident cost calculation defaults to the current project-level employee costs and biz year hours.

Reply:
I think it would make more sense to have them in the incident cost model than the project one, so you have all the settings in one place and you can test different models with different settings including hourly reates, etc. I don't know if there's a place today where you can get individuals' hourly rates.